### PR TITLE
Make TM recover from TO being unavailable at startup

### PR DESCRIPTION
**Pull request by rawlinp**
Always set non-nil TO sessions that can eventually authenticate
themselves when TO recovers. Additionally, instead of exclusively using
one major client version after starting up and logging in, always
attempt the latest major version and fall back to the legacy version in
case of error. This will allow TM to seamlessly transition between using
either major version no matter which order TM and TO are upgraded in.

Closes: #6129

<hr>

## Which Traffic Control components are affected by this PR?

- Traffic Monitor

## What is the best way to verify this PR?

Ensure the added TM unit test passes.

1. Start up TM while TO is also running/available.
2. Once TM has successfully written its backup files to disk, stop TM.
3. In TP, add a new parameter to the TM profile (name = this.is.a.test, config_file = rascal-config.txt, value = test)
4. snapshot the CDN.
5. Shut down TO.
6. Start up TM while TO is still stopped/unavailable.
7. View the log files and observe that TM uses the backup files, continues to operate, and starts polling caches.
8. Start up TO.
9. After one `tm.polling.interval` amount of time (usually 1 minute), TM should successfully poll TO and write the new snapshot backup files. This can be verified by checking the tmconfig.backup file for the parameter created in step 3 (it will be in the `"config": {...}` section).

## If this is a bugfix, which Traffic Control versions contained the bug?

- 5.x
- 6.0.x

## PR submission checklist
- [x] This PR has tests 
- [x] Bug fix, no docs needed
- [x] This PR has a CHANGELOG.md entry 
- [x] This PR **DOES NOT FIX A SERIOUS SECURITY VULNERABILITY** (see [the Apache Software Foundation's security guidelines](https://apache.org/security) for details)

### CODE REVIEW

File: traffic_monitor/handler/handler.go [hunk 1]
```diff
@@ -40,7 +40,6 @@ type OpsConfig struct {
 	HttpsListener string `json:"httpsListener"`
 	CertFile      string `json:"certFile"`
 	KeyFile       string `json:"keyFile"`
-	UsingDummyTO  bool   `json:"usingDummyTO"`
```
[the remark this answers is not in the document]
Comment (rawlinp) [reply traffic_monitor/handler/handler.go:43]: Oh good catch! I didn't even think to look for usages of the json key.

File: traffic_monitor/towrap/towrap.go [hunk 2]
```diff
@@ -36,6 +39,7 @@ import (
 	"github.com/apache/trafficcontrol/traffic_monitor/config"
 	legacyClient "github.com/apache/trafficcontrol/traffic_ops/v2-client"
 	client "github.com/apache/trafficcontrol/traffic_ops/v3-client"
+	"golang.org/x/net/publicsuffix"
 
 	jsoniter "github.com/json-iterator/go"
```
[the remark this answers is not in the document]
Comment (rawlinp) [reply traffic_monitor/towrap/towrap.go:44]: Agreed, will move.

**Review by zrhoffman — APPROVED** [review]
Looks good! Verified that  TM recovers from TO being unavailable on startup.

**Comment by ezkol** [comment]
what if TO restarts after TM is already running  ? will TM recover ? will it attempt to establish a new session with TO ?

**Comment by rob05c** [comment]
> what if TO restarts after TM is already running ? will TM recover ? will it attempt to establish a new session with TO ?

I haven't tested, but that should work. That should automatically be handled by the client: https://github.com/apache/trafficcontrol/blob/master/traffic_ops/toclientlib/toclientlib.go#L576

Any time a request gets a 401 or 403, it tries to login in again once. That handles cookie expirations when it's been too long between requests, but it should also handle if TO is restarted or the secret changed.

**Comment by rob05c** [comment]
Incidentally, all those scenarios--TO unavailable on startup, TO restarted later, cookie expiration--would be great to test with the new TM Integration Test framework.

Though that third might take some hackery with the TO expiration time, for a test that runs quickly.

**Pull request merged**